lib/player_input.rb

```diff
@@ -0,0 +1,43 @@
+require_relative "coordinates"
+require_relative "null_coordinates"
+
+class PlayerInput
+  attr_reader :stdin, :grid, :printer, :coordinates
+  
+  def initialize(stdin, grid, printer)
+    @stdin = stdin
+    @grid = grid
+    @printer = printer
+  end
+
+  def get_valid_coordinate
+    until coordinates.valid?
+      get_coordinate_and_print_error
+    end
+    coordinates.value
+  end
+
+  def get_coordinate_and_print_error
+    get_coordinate_input
+    unless coordinates.valid?
+      printer.print_coordinates_error
+    end
+  end
+
+  def mark_grid
+    grid.mark(get_valid_coordinate)
```

Comment:
I think I already noted that this belongs somewhere else. The name `mark_grid` which is all about another object is the local clue.